### ISSUE TITLE
logalloc_test fails to compile on Ubuntu 15.10 due to __builtin__memcpy_chk() doesn't passes

### ISSUE THREAD

**Issue by syuu1228** [scylladb/scylla#865]
Ubuntu 15.10 fails to compile scylla by following error:
````
FAILED: g++ -MMD -MT build/release/tests/logalloc_test.o -MF build/release/tests/logalloc_test.o.d -std=gnu++1y -g  -Wall -Werror -fvisibility=hidden -pthread -I/home/syuu/scylla/seastar -I/home/syuu/scylla/seastar/build/release/gen  -march=nehalem -DHAVE_DPDK -I/home/syuu/scylla/seastar/build/dpdk/include -march=corei7 -DRTE_MACHINE_CPUFLAG_SSE -DRTE_MACHINE_CPUFLAG_SSE2 -DRTE_MACHINE_CPUFLAG_SSE3 -DRTE_MACHINE_CPUFLAG_SSSE3 -DRTE_MACHINE_CPUFLAG_SSE4_1 -DRTE_MACHINE_CPUFLAG_SSE4_2 -DRTE_COMPILE_TIME_CPUFLAGS=RTE_CPUFLAG_SSE,RTE_CPUFLAG_SSE2,RTE_CPUFLAG_SSE3,RTE_CPUFLAG_SSSE3,RTE_CPUFLAG_SSE4_1,RTE_CPUFLAG_SSE4_2 -Wno-error=literal-suffix -Wno-literal-suffix -Wno-invalid-offsetof -Wno-overloaded-virtual -DHAVE_HWLOC -DHAVE_NUMA  -O2 -I/usr/include/jsoncpp -Wno-maybe-uninitialized  -I. -I build/release/gen -I seastar -I seastar/build/release/gen -c -o build/release/tests/logalloc_test.o tests/logalloc_test.cc
In file included from /usr/include/string.h:639:0,
                 from /usr/include/c++/5/cstring:42,
                 from /usr/include/boost/utility/value_init.hpp:31,
                 from /usr/include/boost/mpl/for_each.hpp:29,
                 from /usr/include/boost/test/unit_test_suite_impl.hpp:29,
                 from /usr/include/boost/test/unit_test_suite.hpp:19,
                 from /usr/include/boost/test/unit_test.hpp:20,
                 from tests/logalloc_test.cc:24:
In function ‘void* memcpy(void*, const void*, size_t)’,
    inlined from ‘managed_bytes::managed_bytes(bytes_view)’ at ./utils/managed_bytes.hh:137:40,
    inlined from ‘managed_bytes::managed_bytes(const bytes&)’ at ./utils/managed_bytes.hh:103:77,
    inlined from ‘test_large_allocation::run_test_case()::<lambda()>::<lambda()>’ at tests/logalloc_test.cc:369:40,
    inlined from ‘decltype(auto) with_allocator(allocation_strategy&, Func&&) [with Func = test_large_allocation::run_test_case()::<lambda()>::<lambda()>]’ at ./utils/allocation_strategy.hh:222:17,
    inlined from ‘test_large_allocation::run_test_case()::<lambda()>’ at tests/logalloc_test.cc:370:19:
/usr/include/x86_64-linux-gnu/bits/string3.h:53:71: error: call to void* __builtin___memcpy_chk(void*, const void*, long unsigned int, long unsigned int) will always overflow destination buffer [-Werror]
   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
                                                                       ^
In function ‘void* memcpy(void*, const void*, size_t)’,
    inlined from ‘managed_bytes::managed_bytes(bytes_view)’ at ./utils/managed_bytes.hh:137:40,
    inlined from ‘managed_bytes::managed_bytes(const bytes&)’ at ./utils/managed_bytes.hh:103:77,
    inlined from ‘test_large_allocation::run_test_case()::<lambda()>::<lambda()>’ at tests/logalloc_test.cc:372:44,
    inlined from ‘decltype(auto) with_allocator(allocation_strategy&, Func&&) [with Func = test_large_allocation::run_test_case()::<lambda()>::<lambda()>]’ at ./utils/allocation_strategy.hh:222:17,
    inlined from ‘test_large_allocation::run_test_case()::<lambda()>’ at tests/logalloc_test.cc:373:19:
/usr/include/x86_64-linux-gnu/bits/string3.h:53:71: error: call to void* __builtin___memcpy_chk(void*, const void*, long unsigned int, long unsigned int) will always overflow destination buffer [-Werror]
   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
                                                                       ^
cc1plus: all warnings being treated as errors
```

It only occurs on logalloc_test, other codes are able to build finely.

**Comment by syuu1228** [comment]
More info: gcc is 5.2.1, boost is 1.58

**Comment by avikivity** [comment]
Maybe we confuse the compiler since we allocate memory, don't initialize it (we promise to with `initialized_later()` but don't keep the promise), then copy it:

```
            while (true) {
                with_allocator(r_evictable.allocator(), [&] {
                    evictable.push_back(bytes(bytes::initialized_later(),element_size));
                });
                with_allocator(r_non_evictable.allocator(), [&] {
                    non_evictable.push_back(bytes(bytes::initialized_later(),element_size));
                });
```

Try clearing the bytes objects before pushing them into the vectors.

**Comment by erichkeane** [comment]
Curious that I wasn't seeing this on my Arch box, which is GCC 5.3.0 and Boost 1.60.  Perhaps this is  something that GCC fixes by 5.3?  I'll try again when I get home.

It seems to me that it is likely confused by the logic in there.  A call is being made to managed_bytes(bytes_view) ctor, which calls the (initialized_later,size_type) ctor.  This then sets the _u object based on the size, either using the blob_storage or small_blob members of the union.  The only way it has of telling this is the external() function (checks vs -1).  I could definitely see the constructor getting confused by that relationship.

**Comment by avikivity** [comment]
Ugh, `atomic_cell::operator atomic_cell_view()` makes this exceedingly hard.
